### PR TITLE
How to Code for Humans

### LINKED ISSUE #27: How to code for humans
Following the ideas of @spreeker mentioned in the old `contrib-guide` and #8.

Good code is written for humans and executable by machines. This means that the human brain understanding your code is just as important as a machine.

## From `contrib-guide`

As a part of #8 

- [x] What is the bar for understanding code and [why](https://phys.org/news/2009-11-brain-magic.html), the Human brain is the most important computer in the mix.
- [x] General guidelines:

> - functions should not contain more then 7 variables.
> - functions can NOT not be longer then 25 Lines Of Code (LOC).
> - functions should be around max ~10-15 lines.
> - DO NOT USE abbreviations. abbreviations make you think. losing precious brain cycles.
> - autocomplete your long and readable_variable_names. do nog use rd_vr_name.
> - **functions should do 1 thing **
> - Write a docstring for every function. if you use AND ... AND split the function up.

### PR DESCRIPTION
**Pull request by bvhme**
Based on the `contrib-guide` (#8) and conversations with @spreeker 

Closes #27.

### CODE REVIEW

File: guides/code-for-humans.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+Here are some basic priciples that can help you write better code.
+
+## Human coding guidelines
+
+### Single Responsibility Principle
+
+* Functions should do 1 thing, and do it well
+* Functions should be around a maximum of 10-15 lines
+* Functions should never exceed 25 lines
+
```
Comment (Vegter) [inline guides/code-for-humans.md]: A comparable argument exists for classes and modules. Describe the responsability of classes and modules, take care that the responsibility is well demarcated, limit dependencies (improves re-usability), etcetera

File: guides/code-for-humans.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+Tip: Describe every function in a _docstring_, if you use “and” in your description you know it does two things and needs to be split up.
+
```
Comment (Vegter) [inline guides/code-for-humans.md]: Or even a little bit more general: Describe functions, classes and modules according to standards like docstrings, jsdoc or whatever standard is applicable to the project

File: guides/code-for-humans.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+### Don’t make me think
+
+When reading code we need all of our brainpower to understand what it is doing, so write clearly and understandable.
+
+* Avoid abbreviations
+* Name functions and methods descriptively
+* Use autocomplete to complete long names
+
+---
```
Comment (Vegter) [inline guides/code-for-humans.md]: Include references to the styleguide to be followed:
- Python PEP8
- Javascript John Papa
- Angular John Papa
- ReactJs ....
- CSS BEM
- etcetera

Keep this section short. Programmers can discuss about this subject for eternity. External style guides are much more effective and... can quite often to be improved using PR's...

I wonder if this whole document could be better replaced by references to standards. What are we doing that is not in a public available standard, and if so, why are we doing this?

**Review by Vegter — CHANGES_REQUESTED** [review]

File: guides/code-for-humans.md (hunk 1 above)
Comment (bvhme) [reply guides/code-for-humans.md]: I've included this. I feel that function scope is included in 'should do 1 thing well'.

File: guides/code-for-humans.md (hunk 2 above)
Comment (bvhme) [reply guides/code-for-humans.md]: I've taken the specific _docstring_ out and replaced in with 'inline documentation'

File: guides/code-for-humans.md (hunk 3 above)
Comment (bvhme) [reply guides/code-for-humans.md]: I've written a tiny blurb about this, although I don't think it is that important

**Comment by bvhme** [comment]
@Vegter I've fixed the issues you've commented on in your review

**Review by Vegter — APPROVED** [review]

**Pull request merged**